src/java_bytecode/java_string_library_preprocess.cpp

```diff
@@ -0,0 +1,2293 @@
+  const typet &type)
+{
+  if(type.id()==ID_pointer)
+  {
+    const pointer_typet &pt=to_pointer_type(type);
+    const typet &subtype=pt.subtype();
+    return is_java_string_buffer_type(subtype);
+  }
+  return false;
+}
+
+/*******************************************************************\
+
+Function: java_string_library_preprocesst::is_java_char_sequence_type
+
+  Inputs: a type
+
+  Output: Boolean telling whether the type is that of java char sequence
+
+\*******************************************************************/
+
+bool java_string_library_preprocesst::is_java_char_sequence_type(
+  const typet &type)
+{
+  return check_java_type(type, "java.lang.CharSequence");
+}
+
+/*******************************************************************\
+
+Function: java_string_library_preprocesst::is_java_char_sequence_pointer_type
+
+  Inputs: a type
+
+  Output: Boolean telling whether the type is that of a pointer
+          to a java char sequence
+
+\*******************************************************************/
+
+bool java_string_library_preprocesst::is_java_char_sequence_pointer_type(
+  const typet &type)
+{
+  if(type.id()==ID_pointer)
+  {
+    const pointer_typet &pt=to_pointer_type(type);
+    const typet &subtype=pt.subtype();
+    return is_java_char_sequence_type(subtype);
+  }
+  return false;
+}
+
+/*******************************************************************\
+
+Function: java_string_library_preprocesst::is_java_char_array_type
+
+  Inputs: a type
+
+  Output: Boolean telling whether the type is that of java char array
+
+\*******************************************************************/
+
+bool java_string_library_preprocesst::is_java_char_array_type(
+  const typet &type)
+{
+  return check_java_type(type, "array[char]");
+}
+
+/*******************************************************************\
+
+Function: java_string_library_preprocesst::is_java_char_array_pointer_type
+
+  Inputs: a type
+
+ Outputs: Boolean telling whether the type is that of a pointer
+          to a java char array
+
+\*******************************************************************/
+
+bool java_string_library_preprocesst::is_java_char_array_pointer_type(
+  const typet &type)
+{
+  if(type.id()==ID_pointer)
+  {
+    const pointer_typet &pt=to_pointer_type(type);
+    const typet &subtype=pt.subtype();
+    return is_java_char_array_type(subtype);
+  }
+  return false;
+}
+
+/*******************************************************************\
+
+Function: java_string_library_preprocesst::string_data_type
+
+  Inputs:
+    symbol_table - a symbol_table containing an entry for java Strings
+
+  Output: the type of data fields in java Strings.
+
+\*******************************************************************/
+
+typet string_data_type(symbol_tablet symbol_table)
+{
+  symbolt sym=symbol_table.lookup("java::java.lang.String");
+  typet concrete_type=sym.type;
+  // TODO: we should look at the name of the field
```

Comment:
It means ideally it should look for field "length" instead of operand #2 (which we hope is "length")